File: rust/crates/wire-mesh-core/src/ports/identity.rs
//! The identity/crypto port. One trait covers the local node's own key
//! (signing), verification against arbitrary issuer keys, and the
//! device-id derivation — so domain logic never imports a crypto library
//! directly.

use wire_mesh_wire::identity::{DeviceId, IdentityKey};

use crate::ports::CoreError;

#[async_trait::async_trait]
pub trait Identity: Send + Sync {
    /// This node's device-id.
    fn device_id(&self) -> &DeviceId;

    /// This node's public identity key.
    fn identity_key(&self) -> &IdentityKey;

    /// Sign `message` (an assembled COSE Sig_structure) with the local
    /// private key.
    async fn sign(&self, message: &[u8]) -> Result<Vec<u8>, CoreError>;

    /// Verify `signature` over `message` against an arbitrary issuer key —
    /// not just the local one, so delegated-token chains verify without
    /// prior contact with any issuer.
    async fn verify(
        &self,
        key: &IdentityKey,
        message: &[u8],
        signature: &[u8],
    ) -> Result<bool, CoreError>;

    /// The device-id derivation rule: SHA-256 of the *raw public-key
    /// bytes*, never of certificate DER (which embeds a serial number and
    /// validity window that change on every reissue).
    fn derive_device_id(&self, public_key: &[u8]) -> DeviceId;

    /// Derives a raw ECDH shared secret against a peer's identity-key --
    /// the asymmetric half of room.rekey's ECIES key-wrapping construction
    /// (wire-mesh#141). Fails closed (rather than panicking or returning
    /// empty bytes) whenever the local key or the peer key is not ES256
    /// (P-256): ECDH is only defined for that curve here, so an
    /// Ed25519-only identity genuinely cannot support it -- the same
    /// "some identities can do this, some can't" reality the TS port
    /// models with an optional method. A caller that needs this and gets
    /// an error must refuse the operation, never substitute a different
    /// construction. The returned bytes are NOT yet an encryption key --
    /// always pass them through HKDF (domain::group_key) first.
    fn derive_shared_secret(&self, _peer_key: &IdentityKey) -> Result<Vec<u8>, CoreError> {
        Err(CoreError::Crypto(
            "this identity cannot derive ECDH shared secrets".to_owned(),
        ))
    }
}

File: ts/packages/core/src/ports/identity.ts
import type { DeviceId, IdentityKey } from "../generated/protocol.js";

/**
 * Wraps the local node's own identity plus the signing/verification primitives capability-token handling needs -- never hardcoded to one crypto library's own API shape (Node's webcrypto, a WASM implementation, a hardware key, etc. can all satisfy this same contract).
 */
export interface IdentityPort {
  readonly deviceId: DeviceId;
  readonly identityKey: IdentityKey;
  /** Signs an already-assembled message (e.g. a COSE Sig_structure) with the local node's own private key. Always a fresh buffer, never a view into shared or offset memory -- directly usable as a CapabilityToken tuple element without further normalisation. */
  sign: (message: Uint8Array) => Promise<Uint8Array<ArrayBuffer>>;
  /** Verifies a signature against an arbitrary identity-key -- not necessarily the local node's own, since capability-token verification checks a token's issuer-key. */
  verify: (
    key: IdentityKey,
    message: Uint8Array,
    signature: Uint8Array,
  ) => Promise<boolean>;
  /** Derives the device-id an arbitrary public key would produce (SHA-256 of the raw public-key bytes -- never a certificate's own DER encoding, the bug both Cascade and agent-comms had to fix). Used to check a token's self-certifying issuer-key against its claimed issuer, not just the local node's own identity. */
  deriveDeviceId: (publicKey: Uint8Array) => Promise<DeviceId>;
  /** Derives a raw ECDH shared secret against a peer's identity-key, the asymmetric half of room.rekey's own ECIES key-wrapping construction (wire-mesh#141) -- HKDF and AES-256-GCM, the symmetric half, need no port since they operate on plain bytes already in hand, with no platform-specific key custody involved. Optional, not a method every implementation must throw from: ECDH is only defined here for an ES256 (P-256) identity-key, so an Ed25519-only identity genuinely cannot support it, the same "some identities can do this, some can't" reality `sign`'s own two-algorithm dispatch already reflects. A caller that needs this and finds it absent must fail closed, never substitute a different construction. Returns the raw shared-secret bytes, NOT yet an encryption key -- always pass through HKDF first, never use ECDH output directly as a symmetric key. */
  deriveSharedSecret?: (
    peerKey: IdentityKey,
  ) => Promise<Uint8Array<ArrayBuffer>>;
}
